File: merge_remaining_prs.py
#!/usr/bin/env python3
"""
Aggressive PR merger to handle remaining open PRs with conflict resolution
"""

import subprocess
import json
import sys
import time

def run_command(cmd, check=True, timeout=60):
    """Run a command and return the result"""
    print(f"→ {cmd}")
    try:
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        if result.stdout:
            print(result.stdout)
        if result.stderr and result.returncode != 0:
            print(f"Error: {result.stderr}")
        return result.returncode == 0, result.stdout.strip()
    except subprocess.TimeoutExpired:
        print(f"⚠ Command timed out")
        return False, ""
    except Exception as e:
        print(f"⚠ Exception: {e}")
        return False, ""

def get_open_prs():
    """Get all open PRs"""
    success, output = run_command("gh pr list --state open --json number,title,headRefName,mergeable")
    if not success:
        return []
    
    try:
        return json.loads(output)
    except json.JSONDecodeError:
        return []

def merge_pr_with_conflict_resolution(pr_number, title, branch):
    """Merge PR with automatic conflict resolution"""
    print(f"\n🔄 Processing PR #{pr_number}: {title}")
    print(f"Branch: {branch}")
    
    # Checkout the PR branch
    success, _ = run_command(f"gh pr checkout {pr_number}", check=False)
    if not success:
        print(f"❌ Failed to checkout PR #{pr_number}")
        return False
    
    # Fetch latest main
    run_command("git fetch origin main", check=False)
    
    # Try to merge main into the branch
    success, _ = run_command("git merge origin/main --no-edit", check=False)
    if not success:
        print("⚠ Merge conflicts detected, resolving automatically...")
        
        # Get list of conflicted files
        success, conflicts = run_command("git diff --name-only --diff-filter=U", check=False)
        if conflicts:
            conflict_files = conflicts.split('\n')
            for file in conflict_files:
                if file.strip():
                    print(f"  Resolving conflicts in {file}...")
                    # Use ours (main branch) for most files
                    run_command(f"git checkout --ours {file}", check=False)
                    run_command(f"git add {file}", check=False)
            
            # Commit the resolved conflicts
            run_command('git commit -m "Resolve merge conflicts automatically"', check=False)
            print("✅ Conflicts resolved")
    
    # Push the updated branch
    run_command(f"git push origin {branch}", check=False)
    
    # Switch back to main
    run_command("git checkout main", check=False)
    
    # Now try to merge the PR
    success, _ = run_command(f"gh pr merge {pr_number} --merge --delete-branch", check=False)
    if success:
        print(f"✅ Successfully merged PR #{pr_number}")
        return True
    else:
        print(f"❌ Failed to merge PR #{pr_number}")
        return False

def main():
    print("🚀 Starting aggressive PR merge process...")
    print("="*70)
    
    # Ensure we're on main and up to date
    run_command("git checkout main", check=False)
    run_command("git pull origin main", check=False)
    
    # Get all open PRs
    prs = get_open_prs()
    print(f"📊 Found {len(prs)} open PRs")
    
    if not prs:
        print("❌ No open PRs found")
        return
    
    # Sort by PR number (oldest first)
    prs.sort(key=lambda x: x['number'])
    
    merged_count = 0
    failed_count = 0
    
    for pr in prs:
        pr_number = pr['number']
        title = pr['title']
        branch = pr['headRefName']
        
        print(f"\n{'='*50}")
        print(f"Processing PR #{pr_number}")
        
        try:
            if merge_pr_with_conflict_resolution(pr_number, title, branch):
                merged_count += 1
            else:
                failed_count += 1
        except Exception as e:
            print(f"❌ Error processing PR #{pr_number}: {e}")
            failed_count += 1
        
        # Small delay between PRs
        time.sleep(1)
    
    print(f"\n{'='*70}")
    print(f"📊 Final Summary:")
    print(f"✅ Successfully merged: {merged_count}")
    print(f"❌ Failed to merge: {failed_count}")
    print(f"📋 Total processed: {len(prs)}")
    
    # Pull latest changes
    print("\n🔄 Pulling latest changes...")
    run_command("git pull origin main", check=False)
    
    print("\n🎉 PR merge process completed!")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n\nFatal error: {e}")
        sys.exit(1)

File: merge_remaining_prs.js
#!/usr/bin/env node

const { execSync } = require('child_process');
const fs = require('fs');

console.log('🚀 Starting Merge Remaining PRs Process...');

function runCommand(command, description) {
  try {
    console.log(`\n🔄 ${description}...`);
    const result = execSync(command, { 
      cwd: '/workspace',
      encoding: 'utf8',
      timeout: 30000
    });
    console.log(`✅ ${description} completed successfully`);
    return result;
  } catch (error) {
    console.log(`⚠️ ${description} had issues: ${error.message}`);
    return null;
  }
}

function main() {
  console.log('📋 Step 1: Checking current status...');
  
  // Check git status
  const status = runCommand('git status --porcelain', 'Checking git status');
  if (status && status.trim()) {
    console.log('⚠️ Working directory has uncommitted changes');
    runCommand('git status --short', 'Showing git status');
  } else {
    console.log('✅ Working directory is clean');
  }
  
  // Ensure we're on main branch
  runCommand('git checkout main', 'Switching to main branch');
  runCommand('git pull origin main', 'Pulling latest main');
  
  console.log('\n📋 Step 2: Fetching all branches...');
  runCommand('git fetch --all', 'Fetching all remote branches');
  
  console.log('\n📋 Step 3: Getting list of remote branches...');
  const branches = runCommand('git branch -r | grep -v "origin/main" | grep -v "origin/HEAD" | sed "s/origin\\///" | tr -d " "', 'Getting remote branches');
  
  if (!branches || !branches.trim()) {
    console.log('✅ No remote branches found to merge. All PRs are already merged!');
    return;
  }
  
  const branchList = branches.trim().split('\n').filter(b => b.trim());
  console.log(`📊 Found ${branchList.length} branches to process`);
  
  if (branchList.length > 20) {
    console.log('📋 First 20 branches:');
    branchList.slice(0, 20).forEach(branch => console.log(`   - ${branch}`));
    console.log(`   ... and ${branchList.length - 20} more`);
  } else {
    console.log('📋 Branches to process:');
    branchList.forEach(branch => console.log(`   - ${branch}`));
  }
  
  console.log('\n📋 Step 4: Processing branches...');
  
  let successful = 0;
  let failed = 0;
  let skipped = 0;
  
  for (let i = 0; i < branchList.length; i++) {
    const branch = branchList[i].trim();
    if (!branch) continue;
    
    console.log(`\n[${i + 1}/${branchList.length}] Processing branch: ${branch}`);
    
    // Check if branch is already merged
    const mergedBranches = runCommand('git branch --merged main', 'Checking merged branches');
    if (mergedBranches && mergedBranches.includes(branch)) {
      console.log(`⏭️ Branch ${branch} is already merged, skipping`);
      skipped++;
      continue;
    }
    
    // Try to merge the branch
    const mergeResult = runCommand(`git merge --no-commit --no-ff "origin/${branch}"`, `Merging ${branch}`);
    
    if (mergeResult !== null) {
      console.log(`✅ Successfully merged ${branch}`);
      runCommand(`git commit -m "Merge ${branch} into main - $(date)"`, `Committing merge of ${branch}`);
      successful++;
    } else {
      console.log(`⚠️ Merge conflicts detected in ${branch}`);
      
      // Try to resolve conflicts automatically
      const conflictFiles = runCommand('git diff --name-only --diff-filter=U', 'Getting conflict files');
      
      if (conflictFiles && conflictFiles.trim()) {
        console.log(`🔧 Resolving conflicts in: ${conflictFiles.trim()}`);
        
        // Resolve conflicts by taking incoming changes
        const files = conflictFiles.trim().split('\n');
        for (const file of files) {
          if (file.trim()) {
            runCommand(`git checkout --theirs "${file.trim()}"`, `Resolving conflicts in ${file}`);
            runCommand(`git add "${file.trim()}"`, `Adding resolved file ${file}`);
          }
        }
        
        // Commit resolved conflicts
        const commitResult = runCommand(`git commit -m "Resolve merge conflicts for ${branch} - $(date)"`, `Committing resolved conflicts for ${branch}`);
        
        if (commitResult !== null) {
          console.log(`✅ Successfully resolved conflicts and merged ${branch}`);
          successful++;
        } else {
          console.log(`❌ Failed to commit resolved conflicts for ${branch}`);
          runCommand('git merge --abort', `Aborting merge of ${branch}`);
          failed++;
        }
      } else {
        console.log(`❌ No conflict files found but merge failed for ${branch}`);
        runCommand('git merge --abort', `Aborting merge of ${branch}`);
        failed++;
      }
    }
    
    // Progress update
    if ((i + 1) % 10 === 0) {
      console.log(`📊 Progress: ${successful} successful, ${failed} failed, ${skipped} skipped`);
    }
  }
  
  // Final summary
  console.log('\n🎉 Merge process completed!');
  console.log('📊 Final Summary:');
  console.log(`   ✅ Successful merges: ${successful}`);
  console.log(`   ❌ Failed merges: ${failed}`);
  console.log(`   ⏭️ Skipped branches: ${skipped}`);
  console.log(`   📈 Total processed: ${branchList.length}`);
  
  // Final status
  console.log('\n📋 Final repository status:');
  runCommand('git status --porcelain', 'Final git status');
  
  console.log('\n🎯 All PRs have been processed!');
}

main();